kong/timing/hooks/http.lua

```diff
@@ -0,0 +1,87 @@
+local _M = {}
+
+local timing
+
+
+local function before_connect_new(self, options)
+  local destination
+  if options.scheme == nil then
+    destination = "unix://" .. options.path
+
+  else
+    local port = options.port or (options.scheme == "http" and 80 or 443)
+    destination = options.scheme .. "://" .. options.host .. ":" .. port
+  end
+
+  self.__kong_timing_destination__ = destination
+
+  timing.enter_context("external_http")
+  timing.enter_context(destination)
+end
+
+
+local function before_connect_deprecated(self, host, port, _options)
+  local destination
+  if type(port) == "number" then
+    destination = "http(s)://" .. host .. ":" .. tostring(port)
+
+  else
+    destination = "unix://" .. host
+  end
+
+  self.__kong_timing_destination__ = destination
+
+  timing.enter_context("external_http")
+  timing.enter_context(destination)
+end
+
+
+local function before_connect(self, arg0, ...)
+  if type(arg0) == "table" then
+    before_connect_new(self, arg0)
+    return
+  end
+
+  before_connect_deprecated(self, arg0, ...)
+end
+
+
+local function after_connect()
+  timing.leave_context() -- leave destination
+  timing.leave_context() -- leave external_http
+end
+
+
+local function before_request(self, _params)
+  timing.enter_context("external_http")
+  timing.enter_context(self.__kong_timing_destination__ or "unknown")
+  timing.enter_context("http_request")
+end
+
+
+local function after_request()
+  timing.leave_context() -- leave http_request
+  timing.leave_context() -- leave destination
+  timing.leave_context() -- leave external_http
+end
+
+
+function _M.register_hooks(timing_module)
+  local http = require("resty.http")
+  local req_dyn_hook = require("kong.dynamic_hook")
+
+  req_dyn_hook.hook_function("timing", http, "connect", 4, {
```

Comment:
Why the child key is `4`? could we add some comments or define a constant?

Reply:
The fourth argument is `max_args`.

Reply:
sorry, it is may mistake. but we should also  explain the reason why it is `4`.

Reply:
This is from the API documents of lua-resty-http, there is a deprecated API: https://github.com/ledgetech/lua-resty-http#TCP-only-connect . It has 4 args including the self.

Reply:
Could you add the comment in code, it will be helpful for future's maintaining work.